src/DependencyInjection/CompilerPass/MiddlewarePass.php

```diff
@@ -100,34 +102,59 @@ private function registerMiddleware(ContainerBuilder $container, array $middlewa
 
             $arguments = $clientDefinition->getArguments();
 
-            $options = [];
-
             if (!empty($arguments)) {
                 $options = array_shift($arguments);
+            } else {
+                $options = [];
             }
 
             if (!isset($options['handler'])) {
-                $handlerStack = new Definition('csa_guzzle.handler_stack');
-                $handlerStack->setFactory(['GuzzleHttp\HandlerStack', 'create']);
+                $handlerStack = new Definition(HandlerStack::class);
+                $handlerStack->setFactory([HandlerStack::class, 'create']);
                 $handlerStack->setPublic(false);
-
-                $clientHandlerStackId = sprintf('csa_guzzle.handler_stack.%s', $clientId);
-
-                $container->setDefinition($clientHandlerStackId, $handlerStack);
-                $options['handler'] = $handlerStack;
+            } else {
+                $handlerStack = $this->wrapHandlerInHandlerStack($options['handler'], $container);
             }
 
-            $handlerStack = $options['handler'];
-
-            foreach ($clientMiddleware as $middleware) {
-                $handlerStack->addMethodCall('push', [new Reference($middleware['id']), $middleware['alias']]);
```

Comment:
This was the problem: handler stack is normally a Reference to another service and not a Definition. So this call was invalid.